### ISSUE TITLE
Support LATERAL subqueries.

### ISSUE THREAD

**Issue by derekmahar** [mithrandie/csvq#24]
Please implement LATERAL subqueries.  According to [_Add LATERAL Joins or CROSS APPLY to Your SQL Tool Chain_](https://blog.jooq.org/2013/12/18/add-lateral-joins-or-cross-apply-to-your-sql-tool-chain/), SQL:1999 introduced "lateral derived tables".  Many popular relational database systems support lateral subqueries:

- [PostgreSQL](https://www.postgresql.org/docs/) [9.3](https://www.postgresql.org/docs/9.3/index.html) introduced [_LATERAL Subqueries_](https://www.postgresql.org/docs/9.3/queries-table-expressions.html#QUERIES-LATERAL).
- [MySQL](https://dev.mysql.com/doc/) [8.0.14](https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-14.html#mysqld-8-0-14-sql-syntax) introduced ["lateral derived tables"](https://dev.mysql.com/doc/refman/8.0/en/lateral-derived-tables.html).
- In Microsoft SQL Server, the join keyword [`APPLY`](https://docs.microsoft.com/en-us/sql/t-sql/queries/from-transact-sql?view=sql-server-ver15#using-apply) implements lateral subqueries.
- Oracle 12c introduced [_LATERAL Inline Views_](https://oracle-base.com/articles/12c/lateral-inline-views-cross-apply-and-outer-apply-joins-12cr1#lateral-inline-views).

In [_PostgreSQL’s LATERAL JOIN_](https://medium.com/kkempin/postgresqls-lateral-join-bfd6bd0199df), [Krzysztof Kempiński](https://medium.com/@kkempin) explains how LATERAL subqueries work in PostgreSQL 9.3.

**Comment by mithrandie** [comment]
Isn’t possible to replace to with, for example, WITH clause and JOIN clause?

**Comment by ondohotola** [comment]
I think for really complicated stuff, just import the CSVs into the SQL database of your choice and have your way with them.

I see CSVQ as a tool to do quick and dirty queries on smallish data sets. And that it does excellently.

el

—

**Comment by mithrandie** [comment]
Yes, I made csvq for simple confirmations and calculations of data. 
Csv has no types, no restrictions, no relations, no indexes, so it is not suitable for large or complicated calculations. And for those reasons, I am not aiming for a complete SQL implementation.

I’ll fix as much as possible obvious bugs or features that are useful in most cases, but in most cases I think that it will work with existing features.
I’m not sure if LATERAL is so needed and there are no other way.

**Comment by derekmahar** [comment]
> Isn’t possible to replace to with, for example, WITH clause and JOIN clause?

Possibly.  I'd have to do more research.  The following solutions are likely inefficient:

- [How can you simulate a LATERAL JOIN in MySQL v8 to execute a subquery or join for each row?](https://stackoverflow.com/q/50563043/107158)
- [Using outer alias in a subquery](https://dba.stackexchange.com/q/195868/83091)

**Comment by derekmahar** [comment]
> Yes, I made csvq for simple confirmations and calculations of data.
> Csv has no types, no restrictions, no relations, no indexes, so it is not suitable for large or complicated calculations. And for those reasons, I am not aiming for a complete SQL implementation.
>
> I’ll fix as much as possible obvious bugs or features that are useful in most cases, but in most cases I think that it will work with existing features.

I highly appreciate that `csvq` already supports much of the SQL standard including advanced syntax like Common Table Expressions and window functions.  I raised this issue for the following reasons:

- `LATERAL` may simplify a query that I wish to execute using `csvq`.
- `LATERAL` is absent from `csvq` and I think it's useful to know in what ways `csvq` deviates from the SQL standard.
- `LATERAL` is part of the SQL standard and I assumed that a goal of `csvq` was to conform to the SQL standard as closely as possible and where it makes sense in the context of CSV files.

I don't expect `csvq` to support all of the SQL standard, but the closer the better.  Over time, you may decide to fill in any gaps that you think you or your users may find useful.

> I’m not sure if LATERAL is so needed and there are no other way.

I'm not yet sure, either, but `LATERAL` does seem to simplify certain queries that use derived tables.

**Comment by derekmahar** [comment]
> I think for really complicated stuff, just import the CSVs into the SQL database of your choice and have your way with them.

I may do this eventually, but for now I'm exploring what computations I may perform using `csvq` alone.

> I see CSVQ as a tool to do quick and dirty queries on smallish data sets. And that it does excellently.

I agree.

**Comment by mithrandie** [comment]
Okey, maybe I will implement it, but now the priority is low because it can be done in other ways.

**Comment by mithrandie** [comment]
LATERAL JOIN is now included and released in version 1.13.0.

**Comment by derekmahar** [comment]
I tested a `LATERAL JOIN` using a slightly modified version of the example in [_PostgreSQL’s LATERAL JOIN_](https://medium.com/kkempin/postgresqls-lateral-join-bfd6bd0199df):

```
$ cat orders.csv | csvq "SELECT * FROM STDIN"
+----+---------+----------------------------+
| id | user_id |         created_at         |
+----+---------+----------------------------+
| 1  | 1       | 2017-06-20 04:35:03.582895 |
| 2  | 2       | 2017-06-20 04:35:07.564973 |
| 3  | 3       | 2017-06-20 04:35:10.986712 |
| 4  | 1       | 2017-06-20 04:58:10.137503 |
| 5  | 3       | 2017-06-20 04:58:17.905277 |
| 6  | 3       | 2017-06-20 04:58:25.289122 |
+----+---------+----------------------------+
```

```sql
-- lateral_query1.sql
SELECT user_id, first_order_time, next_order_time, id FROM
  (SELECT user_id, min(created_at) AS first_order_time FROM orders GROUP BY user_id) o1
  LEFT JOIN LATERAL
  (SELECT id, created_at AS next_order_time
   FROM orders o2
   WHERE o2.user_id = o1.user_id AND created_at > o1.first_order_time
   ORDER BY created_at ASC LIMIT 1)
   o2 ON true;
```

```sql
-- lateral_query2.sql
SELECT user_id, first_order_time, next_order_time, id FROM
  (SELECT user_id, min(created_at) AS first_order_time FROM orders GROUP BY user_id) o1
  INNER JOIN LATERAL
  (SELECT id, created_at AS next_order_time
   FROM orders o2
   WHERE o2.user_id = o1.user_id AND created_at > o1.first_order_time
   ORDER BY created_at ASC LIMIT 1)
   o2 ON true;
```

```
$ cat orders.csv | csvq --source lateral_query1.sql
+---------+----------------------------+-----------------------------+------+
| user_id |      first_order_time      |       next_order_time       |  id  |
+---------+----------------------------+-----------------------------+------+
| 1       | 2017-06-20 04:35:03.582895 | 2017-06-20 04:58:10.137503  | 4    |
| 2       | 2017-06-20 04:35:07.564973 |            NULL             | NULL |
| 3       | 2017-06-20 04:35:10.986712 | 2017-06-20 04:58:17.905277  | 5    |
+---------+----------------------------+-----------------------------+------+
$ cat orders.csv | csvq --source lateral_query2.sql
+---------+----------------------------+-----------------------------+----+
| user_id |      first_order_time      |       next_order_time       | id |
+---------+----------------------------+-----------------------------+----+
| 1       | 2017-06-20 04:35:03.582895 | 2017-06-20 04:58:10.137503  | 4  |
| 3       | 2017-06-20 04:35:10.986712 | 2017-06-20 04:58:17.905277  | 5  |
+---------+----------------------------+-----------------------------+----+
```
The resuls of each these queries matches those in [_PostgreSQL’s LATERAL JOIN_](https://medium.com/kkempin/postgresqls-lateral-join-bfd6bd0199df).
